cmd/syncthing/cli/config.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (C) 2021 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	"github.com/AudriusButkevicius/recli"
+	"github.com/pkg/errors"
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/urfave/cli"
+)
+
+type configHandler struct {
+	original, cfg config.Configuration
+	client        APIClient
+	err           error
+}
+
+func getConfigCommand(f *apiClientFactory) (cli.Command, error) {
+	h := new(configHandler)
+	h.client, h.err = f.getClient()
+	if h.err == nil {
+		h.cfg, h.err = getConfig(h.client)
+	}
+	h.original = h.cfg.Copy()
+
+	// Copy the config and set the default flags
+	recliCfg := recli.DefaultConfig
+	recliCfg.IDTag.Name = "xml"
+	recliCfg.SkipTag.Name = "json"
+
+	commands, err := recli.New(recliCfg).Construct(&h.cfg)
+	if err != nil {
+		return cli.Command{}, fmt.Errorf("config reflect: %w", err)
+	}
+
+	return cli.Command{
+		Name:        "config",
+		HideHelp:    true,
+		Usage:       "Configuration modification command group",
+		Subcommands: commands,
+		After:       h.configAfter,
+	}, nil
+}
+
+func (h *configHandler) configAfter(c *cli.Context) error {
+	if h.err != nil {
+		return h.err
+	}
+	if reflect.DeepEqual(h.cfg, h.original) {
+		return nil
+	}
```

Comment:
If you moved this above the h.err check, --help should not modify the config, in which case we'd end up not returning the error.

Reply:
You need a util.FillNil on the default config it seems, but other than that, seems to work for me.

You can check for "--help" in ctx.Args() apparently, but I feel the DeepEquals early termination is probably less hacky.

Reply:
That's not ok because it means we don't return the error on an actual (read) operation, but just the default value and success.

Reply:
In that case you don't even want to run the command at all.

How about adding a Before handler, if it doesn't have `--help` or `-h` in args, and error is set, just abort the command all together.

And then remove the error check in the After command.

Reply:
Just came to the same conclusion after posting my last comment. However I kept the error check, because `After` is called regardless of whether `Before` returned a non-nil error or not.